### ISSUE TITLE
Support .py (and other language-specific) files

### ISSUE THREAD

**Issue by choldgraf** [amperser/proselint#567]
There are many instances where prose-like writing is embedded within language-specific files. Good examples of this are tutorials for many packages that are out there e.g., [here](https://github.com/mne-tools/mne-python/blob/master/tutorials/plot_background_filtering.py). 

There are probably a subset of `proselint` features that'd be useful in this case (e.g., spellchecker). For example, in Python it could only analyze lines that began with `#` or were embedded within blockquotes, and run a few routines that are appropriate for short lines of text (e.g., line comments).

 I think it'd be useful for creating more polished documentation. Is this something you all have thought about? Something you're friendly to?

**Comment by suchow** [comment]
Yes, this would be great. I think that [SublimeLinter 3](http://www.sublimelinter.com/en/latest/) has this ability baked into it, but I haven't been able to get it to work with Proselint yet. It would be worth another try. It's also possible to bake it into the linter, so that you could for example run `proselint test.py` and have it lint only the docstrings and comments; that'll be some work.

**Comment by michaelpacer** [comment]
This has been on the todo list for a while as a part of the more general goal of inferring formats of files on the basis of more than just the file name extension. We were hoping to set something up to infer that it's a python file, and that 
def blah():
"""…""" 
and # are what need to be linted in that case. The other part of the format specific goal was to use that to automatically exclude some rule subsets that are problematic with particular file formats. But maybe we should split this into two (three?) separate goals as the extraction seems much more straightforward than anything else. Just a regexp, no inference in this case seems to be necessary.

I suppose first step would be to define a scope operation? So the linter can apply to only some lines in a file? I vote we stick to whole line comments for the first pass to keep it simpler.

The other way to do it is to lint the whole file and ignore the lints raised on inappropriate areas. This would make it easier to do comments that come at the end of a code line. But would be substantially less efficient 

But efficiencies can be sacrificed early on and recovered later. So if a post processing filter based on line numbers and column numbers and extracted ranges of valid linting applications would be easier to implement then I'd suggest we go with that. 

Note, one thing that might come up that we should be careful of would be l one range checks that might have some of their inputs in a valid commented area but then the rest of it appears in the code, for which we wouldn't want to trigger a lint if we only want to lint comments.

**Comment by choldgraf** [comment]
That makes sense to me. Maybe just keep it simple at first. Additions along the lines of:

* A checker to determine if the file has the already-supported text-only file types (.md, .txt, etc). If not, then:
  * Have a list of supported code file types, and check if it has that type (e.g., check based on the file extension)
  * For any supported file type, there should be a function that takes the file text and returns an array of indices for lines to keep (begin with `.py` and let people contribute rules for other languages)
  * Lint the whole file using only one module (e.g., spell checker) for now, then return the errors corresponding to lines to keep.

If that works reliably, then it could be a base framework for building on top of, no?

**Comment by michaelpacer** [comment]
Well, part of why it would be good to detect the format of a file independent of its file type is because there can be code blocks in the middle of text files, in which case we would need some way to do the reverse of this here. Fortunately, most coding languages have decently bizarre and specific ways of formatting their code that (from the perspective of natural language parsing) which was why we thought this might be possible. The question is whether it can be done without relying on an "expensive-to-train-every-time-a-new-bug-is-found" neural net based classifier[^fn1]. It would seem to mess with many of the language processing neural nets that consider sequences of words/morphemes/characters and largely ignore punctuation and whitespace. I think that coming up with programmatic rules for this will probably fail, and standard topic models toss away so much information about syntax that I doubt it would work. Maybe a reinforcement learning model with a nonparametric category space could work, since then any new example formats could be incorporated appropriately. Or maybe something like an svn with extracted features from the text would work…

I'm assuming no one has done subdocument splicing into coding formats vs. textual formats even if there is work on identifying programming languages from source text. Some of those approaches will extend more or less well to that task (e.g., I'm guessing svns would fail, since the dimensions inferred would be different depending on where you spliced the document, requiring a different set of features for every possible split, and then there are too many possible splits in any reasonable sized document to make that combinatorics problem go away). 

[fn1]: @choldgraf do you know of anyone who has done programming language identification from textual source files using *any* kind of machine learning algorithm?

**Comment by suchow** [comment]
@michaelpacer GitHub's [linguist](https://github.com/github/linguist) is relevant.

**Comment by choldgraf** [comment]
I think inferring the code type on a block-by-block basis would be more generally useful, but would be an awful lot more work. I don't know the guts of `proselint` but it seems that the former task of choosing lines to keep based on a file extension would be a relatively doable in the short-term, with a more general language-inferrer as a long-term goal.
